File: rightmove-scraper/browser/python/rightmove.py
"""Rightmove scraper using the official Scrapeless Python SDK + Playwright over CDP.
function names + emitted field names match verbatim.

Three surfaces:
- `scrape_properties(urls)`                    — `window.PAGE_MODEL.propertyData`, JMESPath-reduced.
- `find_locations(query)`                      — typeahead API → `"<type>^<id>"` strings.
- `scrape_search(name, id, all, max_props)`    — `/api/property-search/listing/search` paginated.

Uses a GB residential proxy (Rightmove is geo-locked).
"""

from __future__ import annotations

import json
import os
from typing import Any
from urllib.parse import urlencode

import jmespath
from loguru import logger
from parsel import Selector
from playwright.async_api import async_playwright
from scrapeless import Scrapeless
from scrapeless.types import ICreateBrowser

DEFAULT_PROXY_COUNTRY = "GB"
DEFAULT_SESSION_TTL = 180

def _client() -> Scrapeless:
    if not os.environ.get("SCRAPELESS_API_KEY") and not os.environ.get("SCRAPELESS_KEY"):
        raise RuntimeError(
            "SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com"
        )
    if not os.environ.get("SCRAPELESS_API_KEY") and os.environ.get("SCRAPELESS_KEY"):
        os.environ["SCRAPELESS_API_KEY"] = os.environ["SCRAPELESS_KEY"]
    return Scrapeless()

async def _new_browser(proxy_country: str = DEFAULT_PROXY_COUNTRY) -> str:
    client = _client()
    session = client.browser.create(
        ICreateBrowser(proxy_country=proxy_country, session_ttl=DEFAULT_SESSION_TTL)
    )
    return session.browser_ws_endpoint

async def _fetch(
    url: str,
    *,
    as_text: bool = False,
    ready_selector: str | None = None,
    wait_for_global: str | None = None,
    eval_global: str | None = None,
):
    """Fetch a page over CDP.

    Returns:
        - if `eval_global` is set: a tuple `(html, evaluated)` where `evaluated`
          is whatever the JS expression returned (or None on failure).
        - otherwise: the raw HTML (or `document.body.innerText` if `as_text`).
    """
    ws = await _new_browser()
    async with async_playwright() as pw:
        browser = await pw.chromium.connect_over_cdp(ws)
        try:
            page = await browser.new_page()
            await page.goto(url, wait_until="domcontentloaded", timeout=45000)
            if ready_selector:
                try:
                    await page.wait_for_selector(ready_selector, timeout=15000)
                except Exception as e:
                    logger.warning("wait_for_selector failed: {}", e)
            if wait_for_global:
                try:
                    await page.wait_for_function(f"!!window.{wait_for_global}", timeout=15000)
                except Exception as e:
                    logger.warning("wait_for_function({}) failed: {}", wait_for_global, e)
            if as_text:
                body = await page.evaluate("() => document.body && document.body.innerText")
                if body and body.strip():
                    return body
            evaluated = None
            if eval_global:
                try:
                    evaluated = await page.evaluate(eval_global)
                except Exception as e:
                    logger.warning("page.evaluate failed: {}", e)
            html = await page.content()
            if eval_global is not None:
                return html, evaluated
            return html
        finally:
            try:
                await browser.close()
            except Exception:
                pass

# ---------------------------------------------------------------------------
# Parsers — mirror the upstream reference verbatim
# ---------------------------------------------------------------------------

PROPERTY_PARSE_MAP = {
    "id": "id",
    "available": "status.published",
    "archived": "status.archived",
    "phone": "contactInfo.telephoneNumbers.localNumber",
    "bedrooms": "bedrooms",
    "bathrooms": "bathrooms",
    "type": "transactionType",
    "property_type": "propertySubType",
    "tags": "tags",
    "description": "text.description",
    "title": "text.pageTitle",
    "subtitle": "text.propertyPhrase",
    "price": "prices.primaryPrice",
    "price_sqft": "prices.pricePerSqFt",
    "address": "address",
    "latitude": "location.latitude",
    "longitude": "location.longitude",
    "features": "keyFeatures",
    "history": "listingHistory",
    "photos": "images[*].{url: url, caption: caption}",
    "floorplans": "floorplans[*].{url: url, caption: caption}",
    "agency": """customer.{
        id: branchId,
        branch: branchName,
        company: companyName,
        address: displayAddress,
        commercial: commercial,
        buildToRent: buildToRent,
        isNew: isNewHomeDeveloper
    }""",
    "industryAffiliations": "industryAffiliations[*].name",
    "nearest_airports": "nearestAirports[*].{name: name, distance: distance}",
    "nearest_stations": "nearestStations[*].{name: name, distance: distance}",
    "sizings": "sizings[*].{unit: unit, min: minimumSize, max: maximumSize}",
    "brochures": "brochures",
}

def parse_property(data: dict) -> dict:
    """"""
    out: dict[str, Any] = {}
    for key, path in PROPERTY_PARSE_MAP.items():
        out[key] = jmespath.search(path, data)
    return out

def find_json_objects(text: str, decoder: json.JSONDecoder | None = None):
    """Walk a JS payload yielding parsed JSON objects. """
    decoder = decoder or json.JSONDecoder()
    pos = 0
    while True:
        match = text.find("{", pos)
        if match == -1:
            break
        try:
            result, index = decoder.raw_decode(text[match:])
            yield result
            pos = match + index
        except ValueError:
            pos = match + 1

def _revive_devalue(arr: list) -> Any:
    """Revive a Devalue-flattened array (slot 0 is root, numeric values point at
    other slots). Rightmove emits propertyData via this format."""
    seen: dict[int, Any] = {}

    def r(i):
        if not isinstance(i, int) or isinstance(i, bool):
            return i
        if i in seen:
            return seen[i]
        if i < 0 or i >= len(arr):
            return None
        v = arr[i]
        if v is None or not isinstance(v, (dict, list)):
            seen[i] = v
            return v
        if isinstance(v, list):
            out: list = []
            seen[i] = out
            for e in v:
                out.append(r(e) if isinstance(e, int) and not isinstance(e, bool) else e)
            return out
        out_d: dict = {}
        seen[i] = out_d
        for k, val in v.items():
            out_d[k] = r(val) if isinstance(val, int) and not isinstance(val, bool) else val
        return out_d

    return r(0)


def extract_property(html: str) -> dict:
    """Extract `propertyData` from the inline `PAGE_MODEL` JS."""
    sel = Selector(text=html)
    script = sel.xpath("//script[contains(.,'PAGE_MODEL = ')]/text()").get()
    if not script:
        raise RuntimeError("PAGE_MODEL script not found")
    for obj in find_json_objects(script):
        if not isinstance(obj, dict):
            continue
        # New Devalue-encoded form: {"data": "<stringified array>", "encoding": "..."}
        if isinstance(obj.get("data"), str) and "encoding" in obj:
            try:
                arr = json.loads(obj["data"])
            except Exception:
                arr = None
            if isinstance(arr, list) and arr:
                root = _revive_devalue(arr)
                if isinstance(root, dict) and isinstance(root.get("propertyData"), dict):
                    return root["propertyData"]
        # Legacy form: {"propertyData": {...}}
        if isinstance(obj.get("propertyData"), dict):
            return obj["propertyData"]
    raise RuntimeError("propertyData not found in PAGE_MODEL")

# ---------------------------------------------------------------------------
# Scrape functions — mirror the upstream reference verbatim
# ---------------------------------------------------------------------------

async def scrape_properties(urls: list[str]) -> list[dict]:
    """Scrape Rightmove property listings for property data."""
    properties: list[dict] = []
    for url in urls:
        html = await _fetch(
            url,
            ready_selector='#propertyHeader, [data-test="property-header"]',
        )
        logger.info("scraping property page {}", url)
        property_data = extract_property(html)
        if not property_data:
            raise RuntimeError(f"propertyData missing for {url}")
        properties.append(parse_property(property_data))
    return properties

async def find_locations(query: str) -> list[str]:
    """Use Rightmove's typeahead API to find location IDs."""
    url = f"https://los.rightmove.co.uk/typeahead?query={query}&limit=10&exclude=STREET"
    body = await _fetch(url, as_text=True)
    data = json.loads(body)
    return [f"{p['type']}^{p['id']}" for p in data.get("matches", [])]

async def scrape_search(
    location_name: str,
    location_id: str,
    scrape_all_properties: bool,
    max_properties: int = 1000,
) -> list[dict]:
    """Scrape properties from Rightmove's search API. """
    RESULTS_PER_PAGE = 24

    def make_url(offset: int) -> str:
        params = {
            "searchLocation": location_name,
            "useLocationIdentifier": True,
            "locationIdentifier": location_id,
            "radius": "0.0",
            "_includeSSTC": True,
            "index": offset,
            "sortType": "2",
            "channel": "BUY",
            "transactionType": "BUY",
        }
        return "https://www.rightmove.co.uk/api/property-search/listing/search?" + urlencode(params)

    first_body = await _fetch(make_url(0), as_text=True)
    first_data = json.loads(first_body)
    results: list[dict] = list(first_data["properties"])
    total_results = int(str(first_data["resultCount"]).replace(",", ""))

    if not scrape_all_properties and max_properties < total_results:
        MAX = max_properties
    else:
        MAX = total_results

    MAX_API = 1000
    offsets = []
    for offset in range(RESULTS_PER_PAGE, MAX, RESULTS_PER_PAGE):
        if offset >= MAX_API:
            break
        offsets.append(offset)
    logger.info("scraping {} more search pages", len(offsets))
    for offset in offsets:
        try:
            body = await _fetch(make_url(offset), as_text=True)
            data = json.loads(body)
            results.extend(data["properties"])
        except Exception as e:
            logger.warning("failed offset {}: {}", offset, e)
    logger.info("scraped {} properties from location id {}", len(results), location_id)
    return results

def to_dict(obj: Any) -> Any:
    if isinstance(obj, list):
        return [to_dict(x) for x in obj]
    if isinstance(obj, dict):
        return {k: to_dict(v) for k, v in obj.items()}
    return obj

File: rightmove-scraper/browser/nodejs/rightmove.mjs
// Rightmove scraper using the official @scrapeless-ai/sdk + puppeteer-core over CDP.
//
// function names + emitted field names match verbatim.
//
// Three surfaces:
// - `scrapeProperties(urls)`                                — window.PAGE_MODEL.propertyData, JMESPath-reduced.
// - `findLocations(query)`                                  — typeahead API → "<type>^<id>" strings.
// - `scrapeSearch(name, id, scrapeAll, maxProperties)`      — /api/property-search/listing/search paginated.
//
// Uses a GB residential proxy (Rightmove is geo-locked).

import { Scrapeless } from "@scrapeless-ai/sdk";
import puppeteer from "puppeteer-core";
import * as cheerio from "cheerio";
import jmespath from "jmespath";

const DEFAULT_PROXY_COUNTRY = "GB";
const DEFAULT_SESSION_TTL = 180;

function requireKey() {
  const k = process.env.SCRAPELESS_API_KEY ?? process.env.SCRAPELESS_KEY;
  if (!k) throw new Error("SCRAPELESS_API_KEY env var not set. Sign up at https://app.scrapeless.com");
  return k;
}

function client() {
  return new Scrapeless({ apiKey: requireKey() });
}

async function newBrowser(proxyCountry = DEFAULT_PROXY_COUNTRY) {
  const { browserWSEndpoint } = await client().browser.create({
    proxyCountry,
    sessionTTL: DEFAULT_SESSION_TTL,
  });
  return browserWSEndpoint;
}

async function fetchPage(url, { asText = false, readySelector = null, waitForGlobal = null, evalGlobal = null } = {}) {
  const ws = await newBrowser();
  let browser;
  try {
    browser = await puppeteer.connect({ browserWSEndpoint: ws });
    const page = await browser.newPage();
    await page.goto(url, { waitUntil: "domcontentloaded", timeout: 45000 });
    if (readySelector) {
      try { await page.waitForSelector(readySelector, { timeout: 15000 }); } catch (_) {}
    }
    if (waitForGlobal) {
      try {
        await page.waitForFunction(`!!window.${waitForGlobal}`, { timeout: 15000 });
      } catch (_) {}
    }
    if (asText) {
      const body = await page.evaluate(() => document.body && document.body.innerText);
      if (body && body.trim()) return body;
    }
    let evaluated = null;
    if (evalGlobal) {
      try {
        evaluated = await page.evaluate(evalGlobal);
      } catch (_) {}
    }
    const html = await page.content();
    return evalGlobal ? { html, evaluated } : html;
  } finally {
    try { await browser?.close(); } catch (_) {}
  }
}

const PROPERTY_PARSE_MAP = {
  id: "id",
  available: "status.published",
  archived: "status.archived",
  phone: "contactInfo.telephoneNumbers.localNumber",
  bedrooms: "bedrooms",
  bathrooms: "bathrooms",
  type: "transactionType",
  property_type: "propertySubType",
  tags: "tags",
  description: "text.description",
  title: "text.pageTitle",
  subtitle: "text.propertyPhrase",
  price: "prices.primaryPrice",
  price_sqft: "prices.pricePerSqFt",
  address: "address",
  latitude: "location.latitude",
  longitude: "location.longitude",
  features: "keyFeatures",
  history: "listingHistory",
  photos: "images[*].{url: url, caption: caption}",
  floorplans: "floorplans[*].{url: url, caption: caption}",
  agency: `customer.{
    id: branchId,
    branch: branchName,
    company: companyName,
    address: displayAddress,
    commercial: commercial,
    buildToRent: buildToRent,
    isNew: isNewHomeDeveloper
  }`,
  industryAffiliations: "industryAffiliations[*].name",
  nearest_airports: "nearestAirports[*].{name: name, distance: distance}",
  nearest_stations: "nearestStations[*].{name: name, distance: distance}",
  sizings: "sizings[*].{unit: unit, min: minimumSize, max: maximumSize}",
  brochures: "brochures",
};

export function parseProperty(data) {
  const out = {};
  for (const [key, path] of Object.entries(PROPERTY_PARSE_MAP)) {
    out[key] = jmespath.search(data, path);
  }
  return out;
}

// Walk a JS payload yielding parsed JSON objects.
export function* findJsonObjects(text) {
  let pos = 0;
  while (true) {
    const match = text.indexOf("{", pos);
    if (match === -1) break;
    // Try expanding the slice until JSON.parse succeeds at a brace boundary.
    let depth = 0;
    let inString = false;
    let escape = false;
    let end = -1;
    for (let i = match; i < text.length; i++) {
      const c = text[i];
      if (escape) { escape = false; continue; }
      if (c === "\\") { escape = true; continue; }
      if (c === '"') { inString = !inString; continue; }
      if (inString) continue;
      if (c === "{") depth++;
      else if (c === "}") {
        depth--;
        if (depth === 0) { end = i + 1; break; }
      }
    }
    if (end === -1) break;
    try {
      yield JSON.parse(text.slice(match, end));
      pos = end;
    } catch (_) {
      pos = match + 1;
    }
  }
}

// Rightmove now ships PAGE_MODEL as a Devalue-style flat array: each slot is a
// node, and numeric values inside an object/array are slot indices that must be
// resolved against the array. Slot 0 is the root.
function reviveDevalue(arr) {
  const seen = new Map();
  const r = (i) => {
    if (typeof i !== "number") return i;
    if (seen.has(i)) return seen.get(i);
    const v = arr[i];
    if (v === undefined) return undefined;
    if (v === null || typeof v !== "object") {
      seen.set(i, v);
      return v;
    }
    if (Array.isArray(v)) {
      const out = [];
      seen.set(i, out);
      for (const e of v) out.push(typeof e === "number" ? r(e) : e);
      return out;
    }
    const out = {};
    seen.set(i, out);
    for (const [k, val] of Object.entries(v)) {
      out[k] = typeof val === "number" ? r(val) : val;
    }
    return out;
  };
  return r(0);
}

export function extractProperty(html) {
  const $ = cheerio.load(html);
  let script = null;
  $("script").each((_, el) => {
    const txt = $(el).html() ?? "";
    if (txt.includes("PAGE_MODEL = ")) {
      script = txt;
      return false;
    }
  });
  if (!script) throw new Error("PAGE_MODEL script not found");
  for (const obj of findJsonObjects(script)) {
    if (!obj) continue;
    // New Devalue-encoded form: { data: "<stringified array>", encoding: "..." }
    if (typeof obj.data === "string" && Object.prototype.hasOwnProperty.call(obj, "encoding")) {
      try {
        const arr = JSON.parse(obj.data);
        if (Array.isArray(arr) && arr.length > 0) {
          const root = reviveDevalue(arr);
          if (root && root.propertyData) return root.propertyData;
        }
      } catch (_) {}
    }
    // Legacy form: { propertyData: {...} }
    if (obj.propertyData && typeof obj.propertyData === "object") return obj.propertyData;
  }
  throw new Error("propertyData not found in PAGE_MODEL");
}

// ---------------- scrape functions (mirror the upstream reference verbatim) ----------------

export async function scrapeProperties(urls) {
  const out = [];
  for (const url of urls) {
    const html = await fetchPage(url, {
      readySelector: '#propertyHeader, [data-test="property-header"]',
    });
    const propertyData = extractProperty(html);
    if (!propertyData) throw new Error(`propertyData missing for ${url}`);
    out.push(parseProperty(propertyData));
  }
  return out;
}

export async function findLocations(query) {
  const url = `https://los.rightmove.co.uk/typeahead?query=${encodeURIComponent(query)}&limit=10&exclude=STREET`;
  const body = await fetchPage(url, { asText: true });
  const stripped = (body ?? "").trim();
  if (!stripped || !stripped.startsWith("{")) return [];
  let data;
  try { data = JSON.parse(stripped); } catch (_) { return []; }
  return (data.matches ?? []).map((p) => `${p.type}^${p.id}`);
}

export async function scrapeSearch(locationName, locationId, scrapeAllProperties, maxProperties = 1000) {
  const RESULTS_PER_PAGE = 24;
  const MAX_API = 1000;
  const makeUrl = (offset) => {
    const params = new URLSearchParams({
      searchLocation: locationName,
      useLocationIdentifier: "true",
      locationIdentifier: locationId,
      radius: "0.0",
      _includeSSTC: "true",
      index: String(offset),
      sortType: "2",
      channel: "BUY",
      transactionType: "BUY",
    });
    return `https://www.rightmove.co.uk/api/property-search/listing/search?${params.toString()}`;
  };

  const firstBody = await fetchPage(makeUrl(0), { asText: true });
  const firstData = JSON.parse(firstBody);
  const results = [...firstData.properties];
  const totalResults = parseInt(String(firstData.resultCount).replace(/,/g, ""), 10);

  const MAX = !scrapeAllProperties && maxProperties < totalResults ? maxProperties : totalResults;
  const offsets = [];
  for (let offset = RESULTS_PER_PAGE; offset < MAX; offset += RESULTS_PER_PAGE) {
    if (offset >= MAX_API) break;
    offsets.push(offset);
  }
  for (const offset of offsets) {
    try {
      const body = await fetchPage(makeUrl(offset), { asText: true });
      const data = JSON.parse(body);
      results.push(...data.properties);
    } catch (e) {
      console.warn(`failed offset ${offset}: ${e.message}`);
    }
  }
  return results;
}
